### PR TITLE
Check hidden types in dead code

### LINKED ISSUE #99490: MIR dead code analysis affects TAIT inference
Given the code:

```
#![feature(type_alias_impl_trait)]

type Tait = impl Sized;

struct One;
fn one() -> Tait { One }

struct Two<T>(T);
fn two() -> Tait { Two::<()>(todo!()) }
```

If you run the code, you can see that the compiler ends up "successfully" inferencing `Tait = One` and succeeding compilation, even though upon visual inspection, the compiler should be raising an error since `fn two` should cause the conflicting inference `Tait = Two<()>`. 

The key here is the `todo!()` macro, which causes the `SimplifyCfg` MIR pass to remove MIR that actually causes us to make this second inference.

Is this expected? How would we even change this behavior, since it would require us to disentangle TAIT inference from borrowck..?

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #99490

r? @compiler-errors

best reviewed commit by commit

### CODE REVIEW

File: compiler/rustc_borrowck/src/region_infer/opaque_types.rs [hunk 1]
```diff
@@ -496,24 +486,10 @@ impl<'tcx> TypeFolder<'tcx> for ReverseMapper<'tcx> {
             }
         }
 
-        let generics = self.tcx().generics_of(self.key.def_id);
         match self.map.get(&r.into()).map(|k| k.unpack()) {
             Some(GenericArgKind::Lifetime(r1)) => r1,
             Some(u) => panic!("region mapped to unexpected kind: {:?}", u),
             None if self.do_not_error => self.tcx.lifetimes.re_static,
-            None if generics.parent.is_some() => {
-                if let Some(hidden_ty) = self.hidden_ty.take() {
```
Comment (compiler-errors) [inline compiler/rustc_borrowck/src/region_infer/opaque_types.rs:505]: Why did this code exist? What did it achieve?

Comment (compiler-errors) [reply compiler/rustc_borrowck/src/region_infer/opaque_types.rs:505]: Oh, well it was dead code anyways. I should read the commit title.

**Comment by compiler-errors** [comment]
I literally wrote up a comment that had bors r+ but it's on a computer that I don't have access to for a day or two.

Anyways this code makes sense to me and thanks for splitting it up into easily reviewable commits.

@bors r+ rollup=never (just in case we want to bisect to this PR)

**Comment by oli-obk** [comment]
This is a small improvement for everything but the ctfe stress test. The stress test regressed in actual CTFE (though it also added a few additional query calls to `specializes`). The only reason for the longer CTFE time that I can see is that `typeck`'s result now contains more data potentially causing more hashing/comparisons in the `impl Trait` heavy part of the stress test. Either way, I think we should ignore it considering this is now actually checking the right thing, so being more expensive is somewhat expected.

@rustbot label: +perf-regression-triaged

**Pull request merged**